subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,672 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+#include <sys/dlist.h>
+#include <sys/check.h>
+
+#include <device.h>
+#include <init.h>
+
+#include <bluetooth/audio/has.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+
+#include "has_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HAS)
+#define LOG_MODULE_NAME bt_has
+#include "common/log.h"
+
+#define IS_HANDLE_VALID(handle) ((handle) != 0x0000)
+#define IS_PRESET_VALID(preset) ((preset)->index != BT_HAS_PRESET_INDEX_NONE)
+
+#define FEATURES_MASK_HEARING_AID_TYPE (BIT(0) | BIT(1))
+#define FEATURES_MASK_PRESET_SYNCHRONIZATION BIT(2)
+#define FEATURES_MASK_INDEPENDENT_PRESETS BIT(3)
+#define FEATURES_MASK_DYNAMIC_PRESETS BIT(4)
+#define FEATURES_MASK_WRITABLE_PRESETS BIT(5)
+
+static struct bt_has_server has_server;
+
+static ssize_t write_control_point(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				   const void *data, uint16_t len, uint16_t offset, uint8_t flags);
+
+static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			     uint16_t len, uint16_t offset)
+{
+	struct bt_has_server *has = attr->user_data;
+
+	BT_DBG("conn %p attr %p offset %d", conn, attr, offset);
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_LONG);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &has->has.features, 1);
+}
+
+static ssize_t read_active_preset_index(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					void *buf, uint16_t len, uint16_t offset)
+{
+	struct bt_has_server *server = attr->user_data;
+
+	BT_DBG("conn %p attr %p offset %d", conn, attr, offset);
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_ATTRIBUTE_NOT_LONG);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &server->has.active_preset_index, 1);
+}
+
+/* Hearing Access Service GATT Attributes */
+BT_GATT_SERVICE_DEFINE(
+	has_svc, BT_GATT_PRIMARY_SERVICE(BT_UUID_HAS),
+	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_FEATURES, BT_GATT_CHRC_READ, BT_GATT_PERM_READ_ENCRYPT,
+			       read_features, NULL, &has_server),
+	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_CONTROL_POINT,
+			       BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE | BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_WRITE_ENCRYPT, NULL, write_control_point, &has_server),
+	BT_GATT_CHARACTERISTIC(BT_UUID_HAS_ACTIVE_PRESET_INDEX,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, BT_GATT_PERM_READ_ENCRYPT,
+			       read_active_preset_index, NULL, &has_server), );
```

Comment:
Extra space and comma before end parenthesis

Reply:
done